From: Miao Xie <miaox@cn.fujitsu.com>
Date: Tue, 25 Jan 2011 15:46:17 +0800
Patch-mainline: yes
References: FATE#306586
Subject: [PATCH] Btrfs: Don't return acl info when mounting with noacl option

Steps to reproduce:

  # mkfs.btrfs /dev/sda2
  # mount /dev/sda2 /mnt
  # touch /mnt/file0
  # setfacl -m 'u:root:x,g::x,o::x' /mnt/file0
  # umount /mnt
  # mount /dev/sda2 -o noacl /mnt
  # getfacl /mnt/file0
  ...
  user::rw-
  user:root:--x
  group::--x
  mask::--x
  other::--x

The output should be:

  user::rw-
  group::--x
  other::--x

Signed-off-by: Miao Xie <miaox@cn.fujitsu.com>
Signed-off-by: Li Zefan <lizf@cn.fujitsu.com>
Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/acl.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/fs/btrfs/acl.c
+++ b/fs/btrfs/acl.c
@@ -36,6 +36,9 @@ static struct posix_acl *btrfs_get_acl(s
 	char *value = NULL;
 	struct posix_acl *acl;
 
+	if (!IS_POSIXACL(inode))
+		return NULL;
+
 	acl = get_cached_acl(inode, type);
 	if (acl != ACL_NOT_CACHED)
 		return acl;
@@ -83,6 +86,9 @@ static int btrfs_xattr_get_acl(struct in
 	struct posix_acl *acl;
 	int ret = 0;
 
+	if (!IS_POSIXACL(inode))
+		return -EOPNOTSUPP;
+
 	acl = btrfs_get_acl(inode, type);
 
 	if (IS_ERR(acl))

